services/dynamic-sidecar/tests/unit/test_api_workflow_service_metrics.py

```diff
@@ -0,0 +1,454 @@
+    heart_beat_messages = resource_tracking_messages[1:-1]
+    assert len(heart_beat_messages) > 0
+    stop_message = resource_tracking_messages[-1]
+
+    assert isinstance(start_message, RabbitResourceTrackingStartedMessage)
+    for heart_beat_message in heart_beat_messages:
+        assert isinstance(heart_beat_message, RabbitResourceTrackingHeartbeatMessage)
+    assert isinstance(stop_message, RabbitResourceTrackingStoppedMessage)
+    assert stop_message.simcore_platform_status == SimcorePlatformStatus.OK
+
+
+@pytest.mark.parametrize("with_compose_down", [True, False])
+async def test_user_services_fail_to_start(
+    mock_core_rabbitmq: dict[str, AsyncMock],
+    app: FastAPI,
+    httpx_async_client: AsyncClient,
+    client: Client,
+    compose_spec: str,
+    mock_metrics_params: CreateServiceMetricsAdditionalParams,
+    with_compose_down: bool,
+    mock_user_services_fail_to_start: None,
+):
+    with pytest.raises(TaskClientResultError):
+        async with periodic_task_result(
+            client=client,
+            task_id=await _get_task_id_create_service_containers(
+                httpx_async_client, compose_spec, mock_metrics_params
+            ),
+            task_timeout=_CREATE_SERVICE_CONTAINERS_TIMEOUT,
+            status_poll_interval=_FAST_STATUS_POLL,
+        ):
+            ...
+    shared_store: SharedStore = app.state.shared_store
+    assert len(shared_store.container_names) == 0
+
+    if with_compose_down:
+        async with periodic_task_result(
+            client=client,
+            task_id=await _get_task_id_docker_compose_down(httpx_async_client),
+            task_timeout=_CREATE_SERVICE_CONTAINERS_TIMEOUT,
+            status_poll_interval=_FAST_STATUS_POLL,
+        ) as result:
+            assert result is None
+
+    # no messages were sent
+    resource_tracking_messages = _get_resource_tracking_messages(mock_core_rabbitmq)
+    assert len(resource_tracking_messages) == 0
+
+
+async def test_user_services_fail_to_stop_or_save_data(
+    mock_core_rabbitmq: dict[str, AsyncMock],
+    app: FastAPI,
+    httpx_async_client: AsyncClient,
+    client: Client,
+    compose_spec: str,
+    container_names: list[str],
+    mock_metrics_params: CreateServiceMetricsAdditionalParams,
+    mock_user_services_fail_to_stop: None,
+):
+    async with periodic_task_result(
+        client=client,
+        task_id=await _get_task_id_create_service_containers(
+            httpx_async_client, compose_spec, mock_metrics_params
+        ),
+        task_timeout=_CREATE_SERVICE_CONTAINERS_TIMEOUT,
+        status_poll_interval=_FAST_STATUS_POLL,
+    ) as result:
+        assert isinstance(result, list)
+        assert len(result) == len(container_names)
+
+    await _wait_for_containers_to_be_running(app)
+
+    # in case of manual intervention multiple stops will be sent
+    _EXPECTED_STOP_MESSAGES = 4
+    for _ in range(_EXPECTED_STOP_MESSAGES):
+        with pytest.raises(TaskClientResultError):
```

Comment:
Q: Also general question, do you store somewhere the information about last send heartbeat? or it is not needed and you just send it regulary from sidecar? thanks